applications/RANSApplication/custom_strategies/algebraic_flux_corrected_scalar_steady_scheme.h

```diff
@@ -0,0 +1,567 @@
+#include "custom_strategies/relaxed_dof_updater.h"
+#include "rans_application_variables.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+template <class TSparseSpace, class TDenseSpace>
+class AlgebraicFluxCorrectedScalarSteadyScheme : public Scheme<TSparseSpace, TDenseSpace>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(AlgebraicFluxCorrectedScalarSteadyScheme);
+
+    using BaseType = Scheme<TSparseSpace, TDenseSpace>;
+
+    using DofsArrayType = typename BaseType::DofsArrayType;
+
+    using TSystemMatrixType = typename BaseType::TSystemMatrixType;
+
+    using TSystemVectorType = typename BaseType::TSystemVectorType;
+
+    using LocalSystemVectorType = typename BaseType::LocalSystemVectorType;
+
+    using LocalSystemMatrixType = typename BaseType::LocalSystemMatrixType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    AlgebraicFluxCorrectedScalarSteadyScheme(const double RelaxationFactor, const Flags rBoundaryFlags)
+        : BaseType(),
+          mRelaxationFactor(RelaxationFactor),
+          mBoundaryFlags(rBoundaryFlags),
+          mrPeriodicIdVar(Variable<int>::StaticObject())
+    {
+        KRATOS_INFO("AlgebraicFluxCorrectedScalarSteadyScheme")
+            << " Using residual based algebraic flux corrected scheme with "
+               "relaxation "
+               "factor = "
+            << std::scientific << mRelaxationFactor << "\n";
+    }
+
+    AlgebraicFluxCorrectedScalarSteadyScheme(const double RelaxationFactor,
+                                             const Flags rBoundaryFlags,
+                                             const Variable<int>& rPeriodicIdVar)
+        : BaseType(),
+          mRelaxationFactor(RelaxationFactor),
+          mBoundaryFlags(rBoundaryFlags),
+          mrPeriodicIdVar(rPeriodicIdVar)
+    {
+        KRATOS_INFO("AlgebraicFluxCorrectedScalarSteadyScheme")
+            << " Using periodic residual based algebraic flux corrected scheme "
+               "with relaxation "
+               "factor = "
+            << std::scientific << mRelaxationFactor << "\n";
+    }
+
+    ~AlgebraicFluxCorrectedScalarSteadyScheme() override = default;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    void Initialize(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        BaseType::Initialize(rModelPart);
+
+        if (mrPeriodicIdVar != Variable<int>::StaticObject())
+        {
+            const int number_of_conditions = rModelPart.NumberOfConditions();
+#pragma omp parallel for
+            for (int i_cond = 0; i_cond < number_of_conditions; ++i_cond)
+            {
+                const ModelPart::ConditionType& r_condition =
+                    *(rModelPart.ConditionsBegin() + i_cond);
+                if (r_condition.Is(PERIODIC))
+                {
+                    // this only supports 2 noded periodic conditions
+                    KRATOS_ERROR_IF(r_condition.GetGeometry().PointsNumber() != 2)
```

Comment:
Does this mean that this implementation doesn't work in the 3D case?

Reply:
no, it works... But in 3D also periodic conditions should only have 2 nodes. (These periodics consider matching meshes in both periodic surfaces)